### DISCUSSION TITLE
patternProperties & properties

### DISCUSSION THREAD

**Discussion by gantsevdenis** [json-schema-org/community#782] [Specification]
So I am implementing a Draft04 validator for fun, and running it against test suite `https://github.com/json-schema-org/JSON-Schema-Test-Suite`.

When reading the algorithm given in the "Example" section of "additionalProperties, properties and patternProperties" (
https://json-schema.org/draft-04/draft-fge-json-schema-validation-00#rfc.section.5.4.4), we have this sentence:
```
Having collected these three sets, the process is as follows:
    remove from "s" all elements of "p", if any;
    for each regex in "pp", remove all elements of "s" which this regex matches.
```
However, the above mentionned test suit actually tells me otherwise. Given:
```
"schema": {
            "properties": {
                "foo": {"type": "array", "maxItems": 3},
                "bar": {"type": "array"}
            },
            "patternProperties": {"f.o": {"minItems": 2}},
            "additionalProperties": {"type": "integer"}
        }
...
"data": {"foo": []},
```
We have:
```
s = ["foo"]
p = ["foo", "bar"]
pp = ["f.o"]
```
Then:
```
- remove from "s" all elements of "p", if any
s = [] // "foo" was removed

- for each regex in "pp", remove all elements of "s" which this regex matches
s is empty, so nothing applies
```
So in my understanding, the above instance should validate, but according to tests, it doesn't. Can someone please explain this?

**Comment by gregsdennis** [comment]
In this case, `additionalProperties` doesn't enter into the validation at all, really.

The validation is failing because it matches the `patternProperties`, which says the array at `/foo` needs to have a minimum of 2 elements, but it has none.

**Comment by gantsevdenis** [comment]
Okey so because `additionalProperties` doesn't have `boolean value 'false' `, the whole algorithm described above doesn't even apply. Correct?

**Reply by gregsdennis** [reply]
It's not about the subschema of `additionalProperties` in this case because all properties are handled by `properties` and `patternProperties`.

`additionalProperties` only handles properties that aren't handled by the other two.  Since `foo` is handled by both, `additionalProperties` ignores it.

This validation would fail even if `additionalProperties` was removed from the schema.

**Reply by gantsevdenis** [reply]
Okey thanks, that makes pass all the related tests cases :)
